lib/apiservers/engine/backends/container.go

```diff
@@ -345,137 +334,288 @@ func (c *ContainerBackend) ContainerExecResize(eid string, height, width int) er
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		handle, err = c.containerProxy.BindTask(op, handle, eid)
-		if err != nil {
-			return err
+		// success condition
+		if _, success := targetStates[ec.State]; success {
+			op.Debug("Target state reached")
+			return ec, nil
 		}
 
-		// exec doesn't have separate attach path so we will decide whether we need interaction/runblocking or not
-		attach := ec.OpenStdin || ec.OpenStdout || ec.OpenStderr
-		if attach {
-			handle, err = c.containerProxy.BindInteraction(ctx, handle, name, eid)
-			if err != nil {
-				op.Errorf("Failed to initiate interactivity during exec start for container(%s) due to error: %s", id, err)
-				return err
-			}
+		// if we have an explicit error message, return that in preference to the generic
+		// wait state
+		if ec != nil && ec.ProcessConfig != nil && ec.ProcessConfig.ErrorMsg != "" {
+			return ec, errors.New(ec.ProcessConfig.ErrorMsg)
 		}
 
-		if err := c.containerProxy.CommitContainerHandle(ctx, handle, name, 0); err != nil {
-			op.Errorf("Failed to commit handle for container(%s) due to error: %s", id, err)
-			return err
+		// if it's not a wait state then bail with error
+		if _, wait := waitStates[ec.State]; !wait {
+			return ec, fmt.Errorf("state: %s", ec.State)
 		}
 
-		// we need to be able to cancel it
-		taskCtx, cancel := trace.WithCancel(&op, "exec task wait on %s", eid)
-		defer cancel()
+		op.Debug("Waiting for state change")
+		err = c.containerProxy.WaitTask(op, handle, name, eid)
+		if err != nil {
+			return ec, err
+		}
+	}
 
-		// we do not return an error here if this fails. TODO: Investigate what exactly happens on error here...
-		// FIXME: This being in a retry could lead to multiple writes to stdout(?)
-		go func() {
-			defer trace.End(trace.Begin(eid))
+	return nil, op.Err()
+}
 
-			// wait property collector
-			if err := c.containerProxy.WaitTask(taskCtx, id, name, eid); err != nil {
-				op.Errorf("Task wait returned %s, canceling the context", err)
+// ContainerExecStart starts a previously set up exec instance. The
+// std streams are set up.
+func (c *ContainerBackend) ContainerExecStart(ctx context.Context, eid string, stdin io.ReadCloser, stdout, stderr io.Writer) error {
+	op := trace.FromContext(ctx, "exec start")
+	defer trace.End(trace.Begin("", op))
 
-				// we can't return a proper error as we close the streams as soon as AttachStreams returns so we mimic Docker and write to stdout directly
-				// https://github.com/docker/docker/blob/a039ca9affe5fa40c4e029d7aae399b26d433fe9/api/server/router/container/exec.go#L114
-				if stdout != nil {
-					stdout.Write([]byte(err.Error() + "\r\n"))
-				}
-				cancel()
-			}
-		}()
+	// 0. start task (with retry)
+	// 1. If attaching, start attach logic (background)
+	// 2. Wait for container to start or attach to fail
+	//   - on failure return
+	// 3. Generate start event - what should occur on start failure?
```

Comment:
We need an answer to this question - check docker to see what event gets generated in the case of a start failure for an exec.

Reply:
@hickeng I have prepared a quick snippet of a successful start followed by a failed start due to a nonexistent binary. see below:

```
2018-05-29T20:40:48.545386674-07:00 container exec_create: /bin/ls / 362e7859b740f72c409144d4d71f830eb9b12890431e387af55d19ed72b0664b (image=busybox, name=adoring_almeida)
2018-05-29T20:40:48.547870435-07:00 container exec_start: /bin/ls / 362e7859b740f72c409144d4d71f830eb9b12890431e387af55d19ed72b0664b (image=busybox, name=adoring_almeida)
2018-05-29T20:40:58.906500420-07:00 container exec_create: /doesnt/exist  362e7859b740f72c409144d4d71f830eb9b12890431e387af55d19ed72b0664b (image=busybox, name=adoring_almeida)
2018-05-29T20:40:58.908079687-07:00 container exec_start: /doesnt/exist  362e7859b740f72c409144d4d71f830eb9b12890431e387af55d19ed72b0664b (image=busybox, name=adoring_almeida)

Reply:
Looks like the event reporting is the same for my version that I am running which is using the following setup: